sycl/doc/extensions/proposed/sycl_ext_oneapi_graph.asciidoc

```diff
@@ -0,0 +1,1162 @@
+
+While a queue is in the recording state, methods performed on that queue which
+are not command submissions behave as normal. This includes waits, throws, and
+queries on the queue. These are all ignored by the graph system, as opposed to
+throwing an exception when in queue recording mode. This is because otherwise
+there would be no thread safe way for a user to check they could call these
+functions without throwing, as a query about the state of the queue may be
+immediately stale.
+
+=== Storage Lifetimes
+
+The lifetime of any buffer recorded as part of a submission
+to a command graph will be extended in keeping with the common reference
+semantics and buffer synchronization rules in the SYCL specification. It will be
+extended either for the lifetime of the graph (including both modifiable graphs
+and the executable graphs created from them) or until the buffer is no longer
+required by the graph (such as after being replaced through executable graph update).
+
+=== Host Tasks
+
+:host-task: https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#subsec:interfaces.hosttasks
+
+A {host-task}[host task] is a native C++ callable, scheduled according to SYCL
+dependency rules. It is valid to record a host task as part of graph, though it
+may lead to sub-optimal graph performance because a host task node may prevent
+the SYCL runtime from submitting the entire executable `command_graph` to the
+device at once.
+
+Host tasks can be updated as part of <<executable-graph-update, executable graph update>>
+by replacing the whole node with the new callable.
+
+== Examples
+
+[NOTE]
+====
+The examples below demonstrate intended usage of the extension, but may not be
+compatible with the proof-of-concept implementation, as the proof-of-concept
+implementation is currently under development.
+====
+
+Examples for demonstrative purposes only, and may leave out details such as how
+input data is set.
+
+=== Dot Product
+
+[source,c++]
+----
+...
+
+#include <sycl/ext/oneapi/experimental/graph.hpp>
+
+int main() {
+  using namespace sycl::ext::oneapi::experimental;
+
+  const size_t n = 10;
+  float alpha = 1.0f;
+  float beta = 2.0f;
+  float gamma = 3.0f;
+
+  sycl::queue q;
+
+  command_graph g;
+
+  float *x , *y, *z;
+
+  float *dotp = sycl::malloc_shared<float>(1, q);
+
+  // Add commands to the graph to create the following topology.
+  //
+  //  x  y  z
+  //   \ | /
+  //     i
+  //    / \
+  //   a   b
+  //    \ / \
+  //     c  fy
+  //     |
+  //     fx
+
+  auto node_x = g.add_malloc_device(x, n * sizeof(float));
+  auto node_y = g.add_malloc_device(y, n * sizeof(float));
+  auto node_z = g.add_malloc_device(z, n * sizeof(float));
+
+  /* init data on the device */
+  auto node_i = g.add([&](sycl::handler &h) {
+    h.parallel_for(n, [=](sycl::id<1> it){
+      const size_t i = it[0];
+      x[i] = 1.0f;
+      y[i] = 2.0f;
+      z[i] = 3.0f;
+    });
+  }, { property::node::depends_on(node_x, node_y, node_z)});
```

Comment:
Might it be beneficial to decouple edges creation from nodes creation? Something like what's possible in https://en.wikipedia.org/wiki/DOT_(graph_description_language).

Reply:
I have just watched this morning https://www.youtube.com/watch?v=jCnBFjkVuN0 Graph Algorithms and Data Structures in C++20 - Phil Ratzloff & Andrew Lumsdaine - CppCon 2022

Reply:
> We could make use of that in this example code to highlight its existence

Yes, that would be great. I think (unit-)tests are/should be the best examples.